### PR TITLE
[WIP] Retrofit 2 integration

### LINKED ISSUE #484: Migrate to Retrofit 2

### PR DESCRIPTION
**Pull request by tasomaniac**
Retrofit 2 integration is done. We may keep this un-merged because Retrofit 2 is still in beta. Once its released there won't be too much API changes and I think this PR will not change at all. I would like to get this reviewed before hand. 

Logging is removed from Retrofit. We were not using it at all. But I add beautiful logging using OkHttp interceptors for Debug builds. 

Created custom high level Callback abstract class that implements retrofit.Callback.

With this I can log actual network errors to Crashlytics with `Timber.e()` automatically. When there is no internet connection in a device, we are logging these. That's why we have Level 5 network errors in Crashlytics like [this](http://crashes.to/s/40f1de1941b). Those should be logged only with debug log.

Moreover, when there is an internet connection error. It is passed to the caller so that they can show `SnackBar` with appropriate network error.

Fixes #484

### CODE REVIEW

File: app/src/main/java/org/gdg/frisbee/android/fragment/SettingsFragment.java [hunk 1]
```diff
@@ -101,36 +99,29 @@ protected Void doInBackground(Void... voids) {
...
+                                        .enqueue(new Callback<GcmRegistrationResponse>() {
                                             @Override
-                                            public void success(GcmRegistrationResponse gcmRegistrationResponse, retrofit.client.Response response) {
+                                            public void onSuccessResponse(GcmRegistrationResponse response) {
                                                 PrefUtils.setGcmSettings(getActivity(), false, null, null);
                                             }
-
-                                            @Override
-                                            public void failure(RetrofitError error) {
-                                                Timber.e(error, "Fail");
```
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/fragment/SettingsFragment.java]: Error logging like these are removed. All of the errors are now logged automatically. 
It is a network connection or timeout error it only debug logs it. If it is a server error, it error logs it and thus sends it to Crashlytics.

**Comment by tasomaniac** [comment]
Finally this builds :+1: 
I will still wait until Retrofit goes out of beta.

File: app/src/main/java/org/gdg/frisbee/android/api/Callback.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            } catch (IOException e) {
+                Timber.e(e, "Network Error!");
+                failure(e, R.string.server_error);
+            }
+        }
+    }
+
+    @Override
+    public final void onFailure(Throwable t) {
+        Timber.d(t, "Network Failure!");
```
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/api/Callback.java]: This means that there is problem with the connection. This will just send a debug log and show network error message.

File: app/src/main/java/org/gdg/frisbee/android/api/Callback.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        failure(t, R.string.offline_alert);
+    }
+
+    public void failure(Throwable t, @StringRes int errorMessage) {
+    }
+
+    public abstract void success(T response);
```
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/api/Callback.java]: Success is abstract and have to be implemented. Failure is not. It is only be implemented when it is needed.

File: app/src/main/java/org/gdg/frisbee/android/api/GdgXHub.java [hunk 4]
```diff
@@ -21,35 +21,29 @@
 
 public interface GdgXHub {
 
-    @GET("/chapters?perpage=-1")
```
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/api/GdgXHub.java]: No functional change here.

File: app/src/main/java/org/gdg/frisbee/android/appwidget/GdgDashClockExtension.java [hunk 5]
```diff
@@ -60,11 +59,6 @@ public void success(ArrayList<Event> events, retrofit.client.Response response)
                                     .visible(false));
                         }
                     }
-
-                    @Override
-                    public void failure(RetrofitError error) {
-                        Timber.e(error, "Error updating Widget");
```
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/appwidget/GdgDashClockExtension.java]: No need to log manually.

File: app/src/main/java/org/gdg/frisbee/android/common/GdgActivity.java [hunk 6]
```diff
@@ -45,15 +50,21 @@
...
+import butterknife.Bind;
 import butterknife.ButterKnife;
 import timber.log.Timber;
 
 public abstract class GdgActivity extends TrackableActivity implements
         GoogleApiClient.ConnectionCallbacks, GoogleApiClient.OnConnectionFailedListener {
 
+    @Nullable
+    @Bind(R.id.content_frame)
+    FrameLayout mContentLayout;
```
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/common/GdgActivity.java]: Added this View here in the Base Activity. It is present in most of our views. Below in `showError` method, I show the Snackbar if this is present. I show Toast, if this cannot be found.

File: app/src/main/java/org/gdg/frisbee/android/eventseries/GdgEventListFragment.java [hunk 7]
```diff
@@ -138,14 +138,4 @@ private void splitEventsAndAddToAdapter(Collection<? extends SimpleEvent> events
         Collections.reverse(past);
         return new Pair<>(upcoming, past);
     }
-
-    private boolean checkValidCache(Object item) {
```
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/eventseries/GdgEventListFragment.java]: This is moved to base class.

File: app/src/main/java/org/gdg/frisbee/android/pulse/PulseActivity.java [hunk 8]
```diff
@@ -97,7 +88,7 @@ public void onNotFound(String key) {
 
     @Override
     public void onBackPressed() {
-        if (!PulseFragment.GLOBAL.equals(mSpinner.getSelectedItem())) {
+        if (mSpinner != null && !PulseFragment.GLOBAL.equals(mSpinner.getSelectedItem())) {
```
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/pulse/PulseActivity.java]: NullPointerException fix.

File: app/src/main/java/org/gdg/frisbee/android/pulse/PulseFragment.java [hunk 9]
```diff
@@ -156,13 +152,13 @@ public void onPutIntoCache() {
                 }
 
                 @Override
-                public void failure(RetrofitError error) {
-                    if (isAdded()) {
-                        Snackbar snackbar = Snackbar.make(getView(), R.string.fetch_chapters_failed,
```
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/pulse/PulseFragment.java]: The error message here was "Fetching chapters failed". It does not makes sense in Pulse page. Generic Server error is used.

**Comment by tasomaniac** [comment]
Anyone up for a review? :)

File: app/src/main/java/org/gdg/frisbee/android/api/GapiOkTransport.java [hunk 10]
```diff
@@ -101,18 +64,6 @@ protected GapiOkHttpRequest buildRequest(String method, String url) throws IOExc
...
         URL connUrl = new URL(url);
         OkUrlFactory factory = new OkUrlFactory(okHttpClient);
-        SSLContext sslContext;
-        try {
-            sslContext = SSLContext.getInstance("TLS");
-            sslContext.init(null, null, null);
-        } catch (GeneralSecurityException e) {
-            throw new AssertionError(); // The system has no TLS. Just give up.
-        }
-        okHttpClient.setSslSocketFactory(sslContext.getSocketFactory());
```
Comment (friedger) [inline app/src/main/java/org/gdg/frisbee/android/api/GapiOkTransport.java]: why was this removed?

File: app/src/main/java/org/gdg/frisbee/android/api/GdeDirectory.java [hunk 11]
```diff
@@ -2,11 +2,11 @@
...
-import retrofit.http.GET;
+import retrofit2.Call;
+import retrofit2.http.GET;
 
 public interface GdeDirectory {
     
-    @GET("/gde/list")
-    void getDirectory(Callback<GdeList> callback);
+    @GET("gde/list")
+    Call<GdeList> getDirectory();
```
Comment (friedger) [inline app/src/main/java/org/gdg/frisbee/android/api/GdeDirectory.java]: +1

File: app/src/main/java/org/gdg/frisbee/android/api/GroupDirectoryFactory.java [hunk 12]
```diff
@@ -1,44 +1,44 @@
...
-                        }
-                    })
+    private static Retrofit provideRestAdapter() {
+        OkHttpClient.Builder client = App.getInstance().getOkHttpClient().newBuilder();
+        client.addInterceptor(new Interceptor() {
+            @Override
+            public Response intercept(Chain chain) throws IOException {
+
+                Request compressedRequest = chain.request().newBuilder()
+                        .header("User-Agent", "GDG-Frisbee/0.1 (Android)")
```
Comment (friedger) [inline app/src/main/java/org/gdg/frisbee/android/api/GroupDirectoryFactory.java]: maybe update the user agent, or even use the build number of the app

File: app/src/main/java/org/gdg/frisbee/android/common/BaseFragment.java [hunk 13]
```diff
@@ -39,4 +43,16 @@ protected boolean isContextValid() {
         }
         return isContextValid;
     }
+
+    protected void showError(@StringRes final int errorStringRes) {
+        if (isContextValid()) {
+            if (getView() != null) {
```
Comment (friedger) [inline app/src/main/java/org/gdg/frisbee/android/common/BaseFragment.java]: getView could be extracted as local variable

**Comment by friedger** [comment]
Looks good to me.
I was thinking how to improve the use showError and onError methods in the callbacks, but haven't found a good suggestion.

File: app/src/main/java/org/gdg/frisbee/android/api/GapiOkTransport.java (hunk 10 above)
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/api/GapiOkTransport.java]: This is already the default implementation in OkHttp.

File: app/src/main/java/org/gdg/frisbee/android/api/GroupDirectoryFactory.java (hunk 12 above)
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/api/GroupDirectoryFactory.java]: This is the Client for `developers.google.com` which is going to be deprecated in the future. We are using only 2 API calls from here.

File: app/src/main/java/org/gdg/frisbee/android/common/BaseFragment.java (hunk 13 above)
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/common/BaseFragment.java]: Should I do it? It can be exracted in `onCreateView`

**Comment by tasomaniac** [comment]
@friedger thank you for the review. 

I have thought about it a lot. If you check the individual commits, I actually tried couple of iterations. Code wise, it is not perfect. I am open to the suggestions. 

User wise, it became really good. Now `Timber.e` will not send error notifications to Crashlytics related to offline user. Our Crashlytics is full of Network failure Non-Fatal errors. Server errors like 500, 404 etc will still be reported to Crashlytics.

**Comment by tasomaniac** [comment]
I don't understand how this is passed in CI. Proguard configuration should be updated too. It is not compiling locally. Updating the proguard config too now.
